gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+  }
+
+  private void setAllDatasetStatesToVerified() {
+    for (Dataset dataset : this.datasets) {
+      dataset.setState(VERIFIED);
+    }
+  }
+
+  /**
+   * Data completeness verification of a folder should give up if timed out.
+   */
+  private boolean shouldGiveUpVerification() {
+    return this.stopwatch.elapsed(TimeUnit.MINUTES) >= this.dataVerifTimeoutMinutes;
+  }
+
+  private boolean shouldPublishDataIfCannotVerifyCompl() {
+    return this.state.getPropAsBoolean(ConfigurationKeys.COMPACTION_PUBLISH_DATA_IF_CANNOT_VERIFY_COMPLETENESS,
+        ConfigurationKeys.DEFAULT_COMPACTION_PUBLISH_DATA_IF_CANNOT_VERIFY_COMPLETENESS);
+  }
+
+  private void submitCompactionJobsAndWaitForCompletion() throws IOException {
+    LOG.info("Submitting compaction jobs. Number of datasets: " + this.datasets.size());
+
+    boolean allDatasetsCompleted = false;
+    while (!allDatasetsCompleted) {
+      allDatasetsCompleted = true;
+      for (Dataset dataset : this.datasets) {
+        MRCompactorJobRunner jobRunner = MRCompactor.this.jobRunnables.get(dataset);
+
+        if (dataset.state() == VERIFIED || dataset.state() == UNVERIFIED) {
+          allDatasetsCompleted = false;
+
+          // Run compaction for a dataset, if it is not already running or completed
+          if (jobRunner == null || jobRunner.status() == ABORTED) {
+            runCompactionForDataset(dataset, dataset.state() == VERIFIED ? true : false);
+          }
+        } else if (dataset.state() == GIVEN_UP) {
+          if (this.shouldPublishDataIfCannotVerifyCompl) {
+            allDatasetsCompleted = false;
+            if (jobRunner == null || jobRunner.status() == ABORTED) {
+              runCompactionForDataset(dataset, true);
+            } else {
+              jobRunner.proceed();
+            }
+          } else {
+            if (jobRunner != null) {
+              jobRunner.abort();
+            }
+          }
+        }
+      }
+
+      if (shouldGiveUpVerification()) {
+
+        // Data completeness verification timed out. Kill verification jobs and mark all UNVERIFIED datasets
+        // as GIVEN_UP.
+        this.verifier.close();
+
+        for (Dataset dataset : this.datasets) {
+          if (dataset.state() == UNVERIFIED) {
+            dataset.setState(GIVEN_UP);
+          }
+        }
+      }
+
+      if (this.stopwatch.elapsed(TimeUnit.MINUTES) >= this.compactionTimeoutMinutes) {
+
+        // Compaction timed out. Killing all compaction jobs running
+        LOG.error("Compaction timed-out. Killing all running jobs");
+        for (MRCompactorJobRunner jobRunner : MRCompactor.this.jobRunnables.values()) {
+          jobRunner.abort();
+        }
+        break;
+      }
+
+      // Sleep for a few seconds before another round
+      try {
+        Thread.sleep(TimeUnit.SECONDS.toMillis(COMPACTION_JOB_WAIT_INTERVAL_SECONDS));
+      } catch (InterruptedException e) {
+        LOG.error("Interruppted while waiting", e);
```

Comment:
InterruptedExceptions should not be swallowed. Most likely someone is trying to shutdown the service. In that case the code should exit gracefully.